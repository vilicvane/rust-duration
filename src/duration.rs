use syn::{
  LitFloat, LitInt, LitStr, Token,
  parse::{Parse, ParseStream},
};

pub struct DurationMacroInput {
  pub duration: LitStr,
  pub scalar: f64,
}

impl Parse for DurationMacroInput {
  fn parse(input: ParseStream) -> syn::Result<Self> {
    let duration: LitStr = input.parse()?;

    let scalar = if input.is_empty() {
      1.0
    } else {
      let reciprocal = {
        let lookahead = input.lookahead1();

        if lookahead.peek(Token![*]) {
          input.parse::<Token![*]>()?;
          false
        } else if lookahead.peek(Token![/]) {
          input.parse::<Token![/]>()?;
          true
        } else {
          return Err(lookahead.error());
        }
      };

      let number = {
        let lookahead = input.lookahead1();

        if lookahead.peek(LitInt) {
          input.parse::<LitInt>()?.base10_parse::<f64>()?
        } else if lookahead.peek(LitFloat) {
          input.parse::<LitFloat>()?.base10_parse::<f64>()?
        } else {
          return Err(lookahead.error());
        }
      };

      if reciprocal { 1.0 / number } else { number }
    };

    Ok(Self { duration, scalar })
  }
}
