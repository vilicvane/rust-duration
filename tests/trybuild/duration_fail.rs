use std::time::Duration;

use lits::duration;

const _DURATION: Duration = duration!("not-a-duration");

fn main() {}
