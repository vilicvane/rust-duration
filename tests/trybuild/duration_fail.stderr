error: failed to parse duration string: expected number at 0
 --> tests/trybuild/duration_fail.rs:5:39
  |
5 | const _DURATION: Duration = duration!("not-a-duration");
  |                                       ^^^^^^^^^^^^^^^^
